frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,382 @@ pub mod pallet {
 	/// reduction of the lock over time as it diminishes, the account owner must use `vest` or
 	/// `vest_other`.
 	///
 	/// Is a no-op if the amount to be vested is zero.
+	/// NOTE: does not validate schedule params.
 	fn add_vesting_schedule(
 		who: &T::AccountId,
 		locked: BalanceOf<T>,
 		per_block: BalanceOf<T>,
-		starting_block: T::BlockNumber
+		starting_block: T::BlockNumber,
 	) -> DispatchResult {
-		if locked.is_zero() { return Ok(()) }
-		if Vesting::<T>::contains_key(who) {
-			Err(Error::<T>::ExistingVestingSchedule)?
-		}
-		let vesting_schedule = VestingInfo {
-			locked,
-			per_block,
-			starting_block
-		};
-		Vesting::<T>::insert(who, vesting_schedule);
-		// it can't fail, but even if somehow it did, we don't really care.
-		let res = Self::update_lock(who.clone());
-		debug_assert!(res.is_ok());
+		if locked.is_zero() { return Ok(()); }
+
+		let vesting_schedule = VestingInfo::new::<T>(locked, per_block, starting_block);
+		let mut schedules = Self::vesting(who).unwrap_or_default();
+
+		// NOTE: we must push the new schedule so that `report_schedule_updates`
+		// will give the correct new locked amount.
+		ensure!(schedules.try_push(vesting_schedule).is_ok(), Error::<T>::AtMaxVestingSchedules);
+
+		let (schedules, locked_now) =
+			Self::report_schedule_updates(schedules.to_vec(), VestingAction::Passive);
+
+		debug_assert!(
+			locked_now > Zero::zero() && schedules.len() > 0 ||
+				locked_now == Zero::zero() && schedules.len() == 0
+		);
+		Self::write_vesting(&who, schedules)?;
+		Self::write_lock(who, locked_now);
+
+		Ok(())
+	}
+
+	// Ensure we can call `add_vesting_schedule` without error. This should always
+	// be called prior to `add_vesting_schedule`.
+	//
+	// NOTE: expects schedule param validation has been done due to different
+	// scenarios having varying requirements.
+	fn can_add_vesting_schedule(
+		who: &T::AccountId,
+		_locked: BalanceOf<T>,
+		_per_block: BalanceOf<T>,
+		_starting_block: T::BlockNumber,
+	) -> DispatchResult {
+		ensure!(
+			Vesting::<T>::decode_len(who).unwrap_or_default() <
+				T::MaxVestingSchedules::get() as usize,
+			Error::<T>::AtMaxVestingSchedules
+		);
+
 		Ok(())
 	}
 
 	/// Remove a vesting schedule for a given account.
-	fn remove_vesting_schedule(who: &T::AccountId) {
-		Vesting::<T>::remove(who);
-		// it can't fail, but even if somehow it did, we don't really care.
-		let res = Self::update_lock(who.clone());
-		debug_assert!(res.is_ok());
+	fn remove_vesting_schedule(who: &T::AccountId, schedule_index: u32) -> DispatchResult {
+		let schedules = Self::vesting(who).ok_or(Error::<T>::NotVesting)?;
+		let remove_action = VestingAction::Remove(schedule_index as usize);
+
+		let (schedules, locked_now) =
+			Self::report_schedule_updates(schedules.to_vec(), remove_action);
+
+		debug_assert!(
+			locked_now > Zero::zero() && schedules.len() > 0 ||
+				locked_now == Zero::zero() && schedules.len() == 0
+		);
```

Comment:
again, we can consolidate this logic into the `write_vesting` function

Reply:
I think you are just on an older commit - see here https://github.com/paritytech/substrate/pull/9202#discussion_r664238954 && this commit: 4d410f7